### DISCUSSION TITLE
DB convention

### DISCUSSION THREAD

**Discussion by kaos** [jeremyevans/sequel#2179] [General]
Hi!

Newbie to Ruby, and this gem. I have a question about the [DB convention](https://github.com/jeremyevans/sequel?tab=readme-ov-file#label-The+DB+convention) found in the readme.

When declaring my models, it seems to be required that a DB connection exists already. How/when/where do you declare your models if you want the DB connection to not be a module constant, but rather something that you create in a method using some input arguments for the connection string?

```ruby
module A
  class DbStuff
    def initialize
        @db = Sequel.connect(...)
    end
  end

  class MyModule < Sequel::Model # whops.. not possible.
  end
end
```

Can I postpone the model declarations until after the connection has been made somehow? I wouldn't mind if the models end up nested within `DbStuff`, like `DbStuff::MyModule` or something like that.

**Comment by kaos** [comment]
I've been pouring over the docs for a few hours, and there's a lot to take in (more so being new to the Ruby syntax as well) -- just now found a topic related to my question here, so this may become the answer: https://sequel.jeremyevans.net/rdoc/files/doc/code_order_rdoc.html#label-Load+Model+Classes

**Reply by kaos** [reply]
Yep.. managed to solve this with a dynamic require per the docs.
```ruby
  @db = ...
  init_schema
  Dir["#{File.dirname(__FILE__)}/models/*.rb"].each{|f| require f}
```

**Reply by HoneyryderChuck** [reply]
It's best practice to use static vars (instead of ivars) to assign db instances to:

```ruby
DB = ...
# instead of @db = ...
```

All in all, you shouldn't be declaring db instances inside class initializers, if you are using models.

**Reply by kaos** [reply]
Maybe this will all dawn on me when I get there.. but imagine I write a CLI tool, and the options for the connection string is taken as arguments. How do I propagate these options to a static module var during loading? (or how else this is best handled)

**Reply by jeremyevans** [reply]
Wait to define the `DB` constant until the connection string is available, and then load the models after that.

**Reply by kaos** [reply]
Here's where my ruby noob comes in. How do I pass the connection string to a module at load time? ... oh, I may get some ideas just thinking along these lines, thanks.

**Reply by jeremyevans** [reply]
"Pass the connection string to a module" isn't really what's going on.  The question is "How to a dynamically define a constant based on input".  There are two basic ways:

1) Load method.  Here you would set the input to a variable that could be globally referenced, then load a file that defines the constant that references it.  In the example below, I'm using a global variable, but it would be better to use an alternative method (attr_accessor on a constant, for example).

```ruby
  $connection_string = connection_string
  require_relative 'path/to/db/constant'
```

```ruby
  DB = Sequel.connect($connection_string)
```

2) Dynamic constant definition.  Here you can set the constant dynamically.

```ruby
  Object.const_set(:DB, Sequel.connect(connection_string))
```

Most applications using Sequel probably use environment variables or config files (e.g. yaml/json) to store the connection information:

```ruby
  DB = Sequel.connect(ENV['CONNECTION_STRING'])
  DB = Sequel.connect(JSON.parse(File.read('connection_info.json')))
```

If you can do that instead of trying to get the connection string info in one place and then dynamically define the constant, I would recommend it.
